### ISSUE TITLE
[BUG] cuda_async_memory_resource.hpp:77: cudaErrorInvalidValue on WSL2 in Windows10

### ISSUE THREAD

**Issue by gerashegalov** [rapidsai/rmm#1029]
**Describe the bug**
When the Windows Host is on the most recent NVIDIA Driver `512.15`, CUDA Driver `11.6.127` Spark RAPIDS build on CUDA 11.5 runtime in WSL2 VM exhibits the following exception during initialization:
```
ai.rapids.cudf.CudfException: CUDA error at: /home/jenkins/agent/workspace/jenkins-spark-rapids-jni_nightly-dev-github-96-cuda11/thirdparty/cudf/cpp/build/_deps/rmm-src/include/rmm/mr/device/cuda_async_memory_resource.hpp:77: cudaErrorInvalidValue invalid argument
```

<details><summary>exception stacktrace</summary>
<pre>
22/04/29 18:45:00 INFO GpuDeviceManager: Initializing RMM ASYNC pool size = 1536.0 MB on gpuId 0
22/04/29 18:45:00 INFO GpuDeviceManager: Using per-thread default stream
22/04/29 18:45:00 ERROR RapidsExecutorPlugin: Exception in the executor plugin, shutting down!
ai.rapids.cudf.CudfException: CUDA error at: /home/jenkins/agent/workspace/jenkins-spark-rapids-jni_nightly-dev-github-96-cuda11/thirdparty/cudf/cpp/build/_deps/rmm-src/include/rmm/mr/device/cuda_async_memory_resource.hpp:77: cudaErrorInvalidValue invalid argument
        at ai.rapids.cudf.Rmm.initializeInternal(Native Method)
        at ai.rapids.cudf.Rmm.initialize(Rmm.java:119)
        at com.nvidia.spark.rapids.GpuDeviceManager$.initializeRmm(GpuDeviceManager.scala:300)
        at com.nvidia.spark.rapids.GpuDeviceManager$.initializeMemory(GpuDeviceManager.scala:330)
        at com.nvidia.spark.rapids.GpuDeviceManager$.initializeGpuAndMemory(GpuDeviceManager.scala:137)
        at com.nvidia.spark.rapids.RapidsExecutorPlugin.init(Plugin.scala:232)
        at org.apache.spark.internal.plugin.ExecutorPluginContainer.$anonfun$executorPlugins$1(PluginContainer.scala:125)
        at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:293)
        at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
        at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
        at scala.collection.TraversableLike.flatMap(TraversableLike.scala:293)
        at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:290)
        at scala.collection.AbstractTraversable.flatMap(Traversable.scala:108)
        at org.apache.spark.internal.plugin.ExecutorPluginContainer.<init>(PluginContainer.scala:113)
        at org.apache.spark.internal.plugin.PluginContainer$.apply(PluginContainer.scala:211)
        at org.apache.spark.internal.plugin.PluginContainer$.apply(PluginContainer.scala:199)
        at org.apache.spark.executor.Executor.$anonfun$plugins$1(Executor.scala:253)
        at org.apache.spark.util.Utils$.withContextClassLoader(Utils.scala:231)
        at org.apache.spark.executor.Executor.<init>(Executor.scala:253)
        at org.apache.spark.scheduler.local.LocalEndpoint.<init>(LocalSchedulerBackend.scala:64)
        at org.apache.spark.scheduler.local.LocalSchedulerBackend.start(LocalSchedulerBackend.scala:132)
        at org.apache.spark.scheduler.TaskSchedulerImpl.start(TaskSchedulerImpl.scala:220)
        at org.apache.spark.SparkContext.<init>(SparkContext.scala:581)
        at org.apache.spark.api.java.JavaSparkContext.<init>(JavaSparkContext.scala:58)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:247)
        at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
        at py4j.Gateway.invoke(Gateway.java:238)
        at py4j.commands.ConstructorCommand.invokeConstructor(ConstructorCommand.java:80)
        at py4j.commands.ConstructorCommand.execute(ConstructorCommand.java:69)
        at py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)
        at py4j.ClientServerConnection.run(ClientServerConnection.java:106)
        at java.lang.Thread.run(Thread.java:748)
</pre> 
</details>

**Steps/Code to reproduce bug**
1. Create an Ubuntu 20.04 based WSL
2. Clone https://github.com/NVIDIA/spark-rapids
3.  Download and extract Spark 3.2.1 tar ball spark-3.2.1-bin-hadoop3.2
5. Set SPARK_HOME environment variable to the location of local spark-3.2.1-bin-hadoop3.2 directory
6. Install JDK8 and set its root as JAVA_HOME environment variable
7. Install Maven, pytest, ... other 
8.  Run any integration test such e.g. `~/gits/NVIDIA/spark-rapids$ ./integration_tests/run_pyspark_from_build.sh -k array_test` 

**Expected behavior**
The test should succeed as with older driver versions
Driver Version                            : 496.13
CUDA Version                              : 11.5

**Environment details (please complete the following information):**
 - Environment location: WSL2 VM on Windows10 bare-metal host
 - Method of RMM install: statically linked binary with cudfjni, cudf, spark-rapids-jni inside a JAR
  - Please run and attach the output of the `rmm/print_env.sh` script to gather relevant environment details
  
<details><summary>output of the `rmm/print_env.sh`</summary>
**git***
commit 573351d9d0f79cf80e202ed515a0827678043c96 (HEAD -> branch-22.06, origin/branch-22.06, origin/HEAD)
Author: fkallen <37172449+fkallen@users.noreply.github.com>
Date:   Wed Apr 27 01:16:28 2022 +0200

    Add rmm::exec_policy_nosync (#1009)
    
    Add `rmm:exec_policy_nosync`.
    
    For Thrust 1.16 and newer, this policy is based on `execute_on_stream_nosync_base` instead of `execute_on_stream_base` to allow Thrust algorithms to skip synchronizations.
    
    For older Thrust versions, this is an alias to rmm::exec_policy
    
    Authors:
      - https://github.com/fkallen
    
    Approvers:
      - Rong Ou (https://github.com/rongou)
      - Mark Harris (https://github.com/harrism)
      - Jake Hemstad (https://github.com/jrhemstad)
      - Christopher Harris (https://github.com/cwharris)
    
    URL: https://github.com/rapidsai/rmm/pull/1009

***OS Information***
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=20.04
DISTRIB_CODENAME=focal
DISTRIB_DESCRIPTION="Ubuntu 20.04 LTS"
NAME="Ubuntu"
VERSION="20.04 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
Linux NV-3L4YVG3 5.10.102.1-microsoft-standard-WSL2 #1 SMP Wed Mar 2 00:30:59 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

***GPU Information***
Fri Apr 29 12:28:34 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 510.60.02    Driver Version: 512.15       CUDA Version: 11.6     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA RTX A300...  On   | 00000000:01:00.0 Off |                  N/A |
| N/A   49C    P0    20W /  N/A |      0MiB /  6144MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+

***CPU***
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   39 bits physical, 48 bits virtual
CPU(s):                          16
On-line CPU(s) list:             0-15
Thread(s) per core:              2
Core(s) per socket:              8
Socket(s):                       1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           141
Model name:                      11th Gen Intel(R) Core(TM) i7-11850H @ 2.50GHz
Stepping:                        1
CPU MHz:                         2496.001
BogoMIPS:                        4992.00
Hypervisor vendor:               Microsoft
Virtualization type:             full
L1d cache:                       384 KiB
L1i cache:                       256 KiB
L2 cache:                        10 MiB
L3 cache:                        24 MiB
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology tsc_reliable nonstop_tsc cpuid pni pclmulqdq ssse3 f
                                 ma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx
                                  smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx512vbmi umip avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid fsrm avx512_vp2intersect 
                                 flush_l1d arch_capabilities

***CMake***

***g++***
/usr/bin/g++
g++ (Ubuntu 9.3.0-10ubuntu2) 9.3.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

***nvcc***

***Python***
/home/user/miniconda3/bin/python
Python 3.9.7

***Environment Variables***
PATH                            : /home/user/miniconda3/bin:/home/user/miniconda3/condabin:/home/user/.vscode-server-insiders/bin/8632e9c4096eab5773958b9fb3f45f92efba74a8/bin/remote-cli:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/usr/lib/wsl/lib:/mnt/c/Program
Files/NVIDIA                    : GPU
Computing                       : Toolkit/CUDA/v11.5/bin:/mnt/c/Program
Files/NVIDIA                    : GPU
Computing                       : Toolkit/CUDA/v11.5/libnvvp:/mnt/c/Windows/system32:/mnt/c/Windows:/mnt/c/Windows/System32/Wbem:/mnt/c/Windows/System32/WindowsPowerShell/v1.0:/mnt/c/Windows/System32/OpenSSH:/mnt/c/Program
Files/NVIDIA                    : Corporation/NVIDIA
NvDLISR:/mnt/c/Program          : Files/NVIDIA
Corporation/Nsight              : Compute
2021.3.1:/mnt/c/Program         : Files/PowerShell/7:/mnt/c/Program
Files/PuTTY:/mnt/c/Program      : Files/Git/cmd:/mnt/c/Users/user/AppData/Local/Microsoft/WindowsApps:/mnt/c/Users/user/AppData/Local/Programs/Microsoft
VS                              : Code
Insiders/bin                    : 
LD_LIBRARY_PATH                 : 
NUMBAPRO_NVVM                   : 
NUMBAPRO_LIBDEVICE              : 
CONDA_PREFIX                    : /home/user/miniconda3
PYTHON_PATH                     : 

***conda packages***
/home/user/miniconda3/bin/conda
# packages in environment at /home/user/miniconda3:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
_openmp_mutex             4.5                       1_gnu  
brotlipy                  0.7.0           py39h27cfd23_1003  
ca-certificates           2022.3.29            h06a4308_0  
certifi                   2021.10.8        py39h06a4308_2  
cffi                      1.15.0           py39hd667e15_1  
charset-normalizer        2.0.4              pyhd3eb1b0_0  
colorama                  0.4.4              pyhd3eb1b0_0  
conda                     4.12.0           py39h06a4308_0  
conda-package-handling    1.8.1            py39h7f8727e_0  
cryptography              36.0.0           py39h9ce1e76_0  
idna                      3.3                pyhd3eb1b0_0  
ld_impl_linux-64          2.35.1               h7274673_9  
libffi                    3.3                  he6710b0_2  
libgcc-ng                 9.3.0               h5101ec6_17  
libgomp                   9.3.0               h5101ec6_17  
libstdcxx-ng              9.3.0               hd4cf53a_17  
ncurses                   6.3                  h7f8727e_2  
openssl                   1.1.1n               h7f8727e_0  
pip                       21.2.4           py39h06a4308_0  
pycosat                   0.6.3            py39h27cfd23_0  
pycparser                 2.21               pyhd3eb1b0_0  
pyopenssl                 22.0.0             pyhd3eb1b0_0  
pysocks                   1.7.1            py39h06a4308_0  
python                    3.9.7                h12debd9_1  
readline                  8.1.2                h7f8727e_1  
requests                  2.27.1             pyhd3eb1b0_0  
ruamel_yaml               0.15.100         py39h27cfd23_0  
setuptools                58.0.4           py39h06a4308_0  
sqlite                    3.38.2               hc218d9a_0  
tk                        8.6.11               h1ccaba5_0  
tqdm                      4.63.0             pyhd3eb1b0_0  
tzdata                    2022a                hda174b7_0  
urllib3                   1.26.8             pyhd3eb1b0_0  
wheel                     0.37.1             pyhd3eb1b0_0  
xz                        5.2.5                h7b6447c_0  
yaml                      0.2.5                h7b6447c_0  
zlib                      1.2.11               h7f8727e_4  

**Additional context**
</details>

**Comment by gerashegalov** [comment]
https://github.com/rapidsai/rmm/blob/branch-22.06/include/rmm/mr/device/cuda_async_memory_resource.hpp#L77

**Comment by harrism** [comment]
Did this work on earlier CUDA versions (i.e. is it a regression)?

**Comment by gerashegalov** [comment]
yes, it works for `Driver Version: 496.13       CUDA Version: 11.5`

**Comment by jrhemstad** [comment]
Hm, the only thing that stands out as potentially suspicious is the [`cudaMemHandleTypePosixFileDescriptor`](https://github.com/rapidsai/rmm/blob/573351d9d0f79cf80e202ed515a0827678043c96/include/rmm/mr/device/cuda_async_memory_resource.hpp#L73). I'm pretty sure that WSL _is_ considered a POSIX system, but I could be wrong. 

But if this worked on older CUDA versions, then I don't think this is the issue.

**Comment by harrism** [comment]
Jake's suspicion was right. IPC export is not supported on WSL2 yet. So using anything other than `cudaMemHandleTypeNone` is unsupported. 

@gerashegalov does Spark use IPC? If so, this MR can't be used with Spark on WSL2 currently.

I'm going to open a PR to make the MemHandleType an optional parameter of the constructor to allow opting in to IPC support, with it off by default (`cudaMemHandleTypeNone`).

**Comment by jrhemstad** [comment]
> I'm going to open a PR to make the MemHandleType an optional parameter of the constructor to allow opting in to IPC support, with it off by default (`cudaMemHandleTypeNone`).

I suspect Python is unknowingly dependent on the current behavior of IPC support being enabled for UCX. We'll just want to make sure to update it at the Python/Cython layer. @shwina

**Comment by harrism** [comment]
> I suspect Python is unknowingly dependent on the current behavior of IPC support being enabled for UCX.

Trying to parse the above. Do you mean RMM Python is dependent on RMM's current default of enabling IPC? Or RMM Python is dependent on IPC support being enabled for UCX?

**Comment by jrhemstad** [comment]
> Do you mean RMM Python is dependent on RMM's current default of enabling IPC? Or RMM Python is dependent on IPC support being enabled for UCX?

Both. I believe that dask-cudf would today be dependent on RMM implicitly enabling IPC for the `cuda_async_resource` so that it will work with UCX. All we'd need to do is make sure the Python logic for construct the async resource will enable the IPC support.

**Comment by harrism** [comment]
> All we'd need to do is make sure the Python logic for construct the async resource will enable the IPC support.

I think that is also platform dependent, so it should probably be opt-in as well.

**Comment by harrism** [comment]
I have a WIP PR open. Need to figure out the proper Cython changes, but @gerashegalov could  you try out #1030 to see if it fixes your problem with Spark?

**Comment by gerashegalov** [comment]
> I have a WIP PR open. Need to figure out the proper Cython changes, but @gerashegalov could you try out #1030 to see if it fixes your problem with Spark?

Upgrading / Downgrading the driver on Windows is time consuming. I'll try to get to it today

**Comment by gerashegalov** [comment]
> I have a WIP PR open. Need to figure out the proper Cython changes, but @gerashegalov could you try out #1030 to see if it fixes your problem with Spark?

@harrism I confirmed your fix from #1030 
the easiest reproduction is via cuDF repo
```
# fails
$ ./build.sh  cudfjar

# works
$ cd cpp/build/_deps/rmm-src/
$ curl https://patch-diff.githubusercontent.com/raw/rapidsai/rmm/pull/1030.patch  | patch -p1
$ ./build.sh  cudfjar
```

**Comment by harrism** [comment]
Fixed by #1051
